lib/Monads/trace/Trace_Monad.thy

```diff
@@ -572,139 +460,141 @@ lemma "doE x \<leftarrow> returnOk 1;
   by simp
 
 
+subsection "Interference command"
 
-section "Library of additional Monadic Functions and Combinators"
+text \<open>
+  Interference commands must be inserted in between actions that can be interfered with by
+  commands running in other threads.\<close>
 
+definition last_st_tr :: "(tmid * 's) list \<Rightarrow> 's \<Rightarrow> 's" where
+  "last_st_tr tr s0 \<equiv> hd (map snd tr @ [s0])"
 
-text \<open> Lifting a normal function into the monad type: \<close>
-definition
-  liftM :: "('a \<Rightarrow> 'b) \<Rightarrow> ('s,'a) tmonad \<Rightarrow> ('s, 'b) tmonad"
-where
+definition env_steps :: "('s,unit) tmonad" where
+  "env_steps \<equiv>
+  do
+    s \<leftarrow> get;
+    \<comment> \<open>Add unfiltered environment events to the trace\<close>
+    xs \<leftarrow> select UNIV;
+    tr \<leftarrow> return (map (Pair Env) xs);
+    put_trace tr;
+    \<comment> \<open>Pick the last event of the trace as the final state\<close>
+    put (last_st_tr tr s)
+  od"
+
+definition commit_step :: "('s,unit) tmonad" where
+  "commit_step \<equiv>
+  do
+    s \<leftarrow> get;
+    put_trace [(Me,s)]
+  od"
+
+definition interference :: "('s,unit) tmonad" where
+  "interference \<equiv>
+  do
+    commit_step;
+    env_steps
+  od"
```

Comment:
I'm reviewing commit-by-commit, so this might be addressed in further commits, but for these 3 defs, the `do ... od` indentation isn't right and I'm looking at the `restyle Trace_Monad.thy` commit. Bit confused.

Reply:
Good catch, I missed these. I'll probably put this in the fourth commit though. I haven't been perfect with it but my intention for the first one was to be restyling updates taken from the nondet monad with changes to trace monad sections being left for later.